File: test-suite/go/tests/minio.go
package tests

import (
	"bytes"
	"context"
	"fmt"

	"github.com/google/uuid"
	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

const (
	minioService = "MinIO"
	minioClient  = "go-minio"
)

// TestMinIO tests MinIO connectivity
func TestMinIO(utils TestUtils) ServiceResult {
	host := utils.GetEnv("MINIO_HOST", "minio")
	port := utils.GetEnvInt("MINIO_PORT", 9000)
	accessKey := utils.GetEnv("MINIO_ACCESS_KEY", "minio")
	secretKey := utils.GetEnv("MINIO_SECRET_KEY", "minio12345")

	if gate := utils.VerifyServiceGate(minioService, minioClient, "ENABLE_MINIO", true, host, port); gate != nil {
		return *gate
	}

	return utils.RunCheck(minioService, minioClient, func() (string, error) {
		ctx := context.Background()

		client, err := minio.New(fmt.Sprintf("%s:%d", host, port), &minio.Options{
			Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
			Secure: false,
		})
		if err != nil {
			return "", err
		}

		// Create bucket
		bucketName := fmt.Sprintf("health-check-%s", uuid.New().String())
		err = client.MakeBucket(ctx, bucketName, minio.MakeBucketOptions{})
		if err != nil {
			return "", err
		}

		// Upload object
		objectName := fmt.Sprintf("test-%s.txt", uuid.New().String())
		content := []byte(minioClient)
		reader := bytes.NewReader(content)

		_, err = client.PutObject(ctx, bucketName, objectName, reader, int64(len(content)), minio.PutObjectOptions{
			ContentType: "text/plain",
		})
		if err != nil {
			return "", err
		}

		return fmt.Sprintf("Uploaded object %s to bucket %s", objectName, bucketName), nil
	})
}

File: test-suite/rust/src/tests/minio.rs
use crate::utils::*;
use s3::bucket::Bucket;
use s3::creds::Credentials;
use s3::region::Region;
use uuid::Uuid;

const SERVICE: &str = "MinIO";
const CLIENT_NAME: &str = "rust-s3";

pub fn test_minio() -> ServiceResult {
    let host = get_env("MINIO_HOST", "minio");
    let port = get_env_u16("MINIO_PORT", 9000);
    let access_key = get_env("MINIO_ACCESS_KEY", "minio");
    let secret_key = get_env("MINIO_SECRET_KEY", "minio12345");

    if let Some(gate) = verify_service_gate(SERVICE, CLIENT_NAME, "ENABLE_MINIO", true, &host, port) {
        return gate;
    }

    run_check(SERVICE, CLIENT_NAME, || {
        let endpoint = format!("http://{}:{}", host, port);
        let region = Region::Custom {
            region: "us-east-1".to_string(),
            endpoint,
        };

        let credentials = Credentials::new(
            Some(&access_key),
            Some(&secret_key),
            None,
            None,
            None,
        )?;

        // Use Tokio runtime for reqwest-based clients
        let runtime = tokio::runtime::Runtime::new()?;
        runtime.block_on(async {
            // Create bucket
            let bucket_name = format!("health-check-{}", Uuid::new_v4());
            let create_response = Bucket::create_with_path_style(
                &bucket_name,
                region.clone(),
                credentials.clone(),
                s3::BucketConfiguration::default(),
            )
            .await
            .map_err(|e| format!("Failed to create bucket: {:?}", e))?;

            let bucket = *create_response.bucket;

            // Upload object
            let object_name = format!("test-{}.txt", Uuid::new_v4());
            let content = CLIENT_NAME.as_bytes();

            let response = bucket.put_object(&object_name, content)
                .await
                .map_err(|e| format!("Failed to upload object: {:?}", e))?;

            if response.status_code() != 200 {
                return Err(format!("Upload failed with status: {}", response.status_code()).into());
            }

            Ok(format!(
                "Uploaded object {} to bucket {}",
                object_name, bucket_name
            ))
        })
    })
}
